rswag-specs/lib/rswag/specs/request_factory.rb

```diff
@@ -101,8 +101,27 @@ def add_verb(request, metadata)
         request[:verb] = metadata[:operation][:verb]
       end
 
+      def base_path_from_servers(swagger_doc, use_server = :default)
+        return '' if swagger_doc[:servers].nil? || swagger_doc[:servers].empty?
+        base_paths = swagger_doc[:servers].map do |server|
+          variables = {}
+          server.fetch(:variables, {}).each_pair { |k,v| variables[k] = v[use_server] }
+          server[:url].gsub(/\{(.*?)\}/) { variables[$1.to_sym] }
+        end
+        URI(base_paths.first).path
+      end
+
       def add_path(request, metadata, swagger_doc, parameters, example)
-        template = (swagger_doc[:basePath] || '') + metadata[:path_item][:template]
+        if doc_version(swagger_doc).start_with?('3')
+          if !swagger_doc[:basePath].nil?
+            ActiveSupport::Deprecation.warn('Rswag::Specs: WARNING: basePath is replaced in OpenAPI3! Update your swagger_helper.rb')
+            template = (swagger_doc[:basePath] || '') + metadata[:path_item][:template]
+          else
+            template = base_path_from_servers(swagger_doc) + metadata[:path_item][:template]
+          end
+        else # OpenAPI 2
+          template = (swagger_doc[:basePath] || '') + metadata[:path_item][:template]
+        end
```

Comment:
If not very good at understanding nested conditionals so I rearranged the code to help myself understand.
Your call on which version you'd prefer to add since you did the heavy lifting though :slightly_smiling_face: 

```suggestion
        is_swagger3 = doc_version(swagger_doc).start_with?('3')
        uses_swagger2 = !swagger_doc[:basePath].nil?
        base_path = if uses_swagger2
                      ActiveSupport::Deprecation.warn(<<~DEPRECATION) if is_swagger3
                        Rswag::Specs: WARNING: basePath is replaced in OpenAPI3! Update your swagger_helper.rb
                      DEPRECATION
                      (swagger_doc[:basePath] || '')
                    else
                      base_path_from_servers(swagger_doc)
                    end
        template = base_path + metadata[:path_item][:template]
```

Reply:
I've tried to reduce the complexity a bit, see if it's better :wink: